### PR TITLE
feat: Add method to add analyzer rules to SessionContext

### LINKED ISSUE #10846: API to add analyzer_rules
### Is your feature request related to a problem or challenge?

@goldmedal  reports on discord: https://discord.com/channels/885562378132000778/1166447479609376850/1249052209388326922

Hi, I tried to implement some AnalyzerRule to rewrite and analyze the logical plan. I have implemented the AnalyzerRules that work fine with Analyzer::with_rules. However, when I tried to register them to the SessionContext, I couldn't find any method for it. I can only do something like:

```rust
let ctx = SessionContext::new();
let new_st = ctx.state().add_analyzer_rule(Arc::new(ModelAnalyzeRule::new(Arc::clone(&analyzed_mdl))));
let new_st = new_st.add_analyzer_rule(Arc::new(ModelGenerationRule::new(Arc::clone(&analyzed_mdl))));
let new_ctx = SessionContext::new_with_state(new_st);
```

The new session context works fine with the rules I added, but I can't find a way to register the rules to the session context directly.

Does this approach make sense? Or is there a better way to do this?

Is there any method like SessionContext::register_analyzer_rule that can add the rules to the analyzer rules list?

### Describe the solution you'd like

I would like a function on `SessionContext` that can update the analyzer rules directly

### Describe alternatives you've considered

I suggest adding `SessionContext::add_analyzer_rule` that simply calls `SessionState::add_analyzer_rule`

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by pingsutw**
## Which issue does this PR close?
Closes #10846

## Rationale for this change
The new session context works fine with the rules I added, but there is no way to register the rules to the session context directly.

## What changes are included in this PR?
adding SessionContext::add_analyzer_rule that simply calls SessionState::add_analyzer_rule

## Are these changes tested?
unit tests

### CODE REVIEW

**Review by alamb — COMMENTED** [review]
Thank you for this contribution @pingsutw  🙏 

I was going to say that we should add tests for this API, but it seems as though there are no existing tests for `add_analyzer_rule` 😢 


https://github.com/search?q=repo%3Aapache%2Fdatafusion+add_analyzer_rule&type=code

It is probably time to add an example or test, similar to the user_defined_plan here:
https://github.com/apache/datafusion/blob/3773fb7fb54419f889e7d18b73e9eb48069eb08e/datafusion/core/tests/user_defined/user_defined_plan.rs#L251

However I don't think you have to do it in this PR unless you want to

I filed https://github.com/apache/datafusion/issues/10855 to track the idea

File: datafusion/core/src/execution/context/mod.rs [hunk 1]
```diff
@@ -331,6 +332,15 @@ impl SessionContext {
         self
     }
 
+    /// Adds an analyzer rule to the `SessionState` in the current `SessionContext`.
```
Comment (comphead) [inline datafusion/core/src/execution/context/mod.rs:335]: Would be nice to have an examples, or doc test for the this method

Comment (alamb) [reply datafusion/core/src/execution/context/mod.rs:335]: I agree --  we are tracking adding an example for how to use custom analyzer rules in https://github.com/apache/datafusion/issues/10855, so perhaps we can add the example as part of that ticket (i think @goldmedal  said he may have some time to work on that eventually)

**Comment by alamb** [comment]
@pingsutw  please let me know if you have time to add an example for this API, otherwise I think perhaps we can merge it in and add coverage as part of https://github.com/apache/datafusion/issues/10855 (or we could hold off on this code until the PR that adds the example)

**Comment by pingsutw** [comment]
I'll update my pr and add an example tonight

File: datafusion/core/src/execution/session_state.rs [hunk 2]
```diff
@@ -387,9 +387,9 @@ impl SessionState {
     /// Add `analyzer_rule` to the end of the list of
     /// [`AnalyzerRule`]s used to rewrite queries.
     pub fn add_analyzer_rule(
-        mut self,
+        &mut self,
```
[the remark this answers is not in the document]
Comment (pingsutw) [reply datafusion/core/src/execution/session_state.rs:390]: Should we also update `add_optimizer_rule` `add_physical_optimizer_rule`? (mut self -> &mut self)

Also, do we need to add `with_optimizer_rule` and `with_physical_optimizer_rule` to make it consistent?
If so, I can do it in a separate PR.

**Review by alamb — COMMENTED** [review]
Thank you very much @pingsutw  -- this is looking good

File: datafusion/core/src/execution/context/mod.rs (hunk 1 above)
Comment (alamb) [reply datafusion/core/src/execution/context/mod.rs:335]: FWIW I have a WIP PR to improve these examples -- I hope to get it up for review sometime this weekend

**Review by alamb — APPROVED** [review]
Looks good to me -- thank you @pingsutw 

I merged this PR up from main to resolve a conflict

**Comment by pingsutw** [comment]
> I merged this PR up from main to resolve a conflict

Thank you for the help!

**Pull request merged**